src/test/spec/unified_runner/matcher.rs

```diff
@@ -78,21 +104,75 @@ pub fn events_match(
             }
         }
         (
-            TestEvent::Failed {
+            ExpectedCommandEvent::Failed {
                 command_name: actual_command_name,
+                has_service_id: actual_has_service_id,
             },
-            TestEvent::Failed {
+            ExpectedCommandEvent::Failed {
                 command_name: expected_command_name,
+                has_service_id: expected_has_service_id,
             },
-        ) => match (expected_command_name, actual_command_name) {
-            (Some(expected), Some(actual)) => expected == actual,
-            (Some(_), None) => false,
-            _ => true,
-        },
+        ) => {
+            if expected_has_service_id.is_some() && actual_has_service_id != expected_has_service_id
+            {
+                return false;
+            }
+            match (expected_command_name, actual_command_name) {
+                (Some(expected), Some(actual)) => expected == actual,
+                (Some(_), None) => false,
+                _ => true,
+            }
+        }
         _ => false,
     }
 }
 
+fn cmap_events_match(actual: &ExpectedCmapEvent, expected: &ExpectedCmapEvent) -> bool {
+    match (actual, expected) {
+        (
+            ExpectedCmapEvent::PoolCleared {
+                has_service_id: expected_has_service_id,
+            },
+            ExpectedCmapEvent::PoolCleared {
+                has_service_id: actual_has_service_id,
+            },
+        ) => {
+            if expected_has_service_id.is_some() && actual_has_service_id != expected_has_service_id
+            {
+                return false;
+            }
+            true
+        }
+        (
+            ExpectedCmapEvent::ConnectionClosed {
+                reason: expected_reason,
+            },
+            ExpectedCmapEvent::ConnectionClosed {
+                reason: actual_reason,
+            },
+        ) => {
+            if expected_reason.is_some() && actual_reason != expected_reason {
+                return false;
+            }
+            true
+        }
+        (
+            ExpectedCmapEvent::ConnectionCheckOutFailed {
+                reason: expected_reason,
+            },
+            ExpectedCmapEvent::ConnectionCheckOutFailed {
+                reason: actual_reason,
+            },
+        ) => {
+            if expected_reason.is_some() && actual_reason != expected_reason {
+                return false;
+            }
+            true
+        }
+        _ => actual == expected,
```

Comment:
The rest of the `ExpectedCmapEvent` variants have zero fields, so `==` is conveniently applicable.